src/material/core/datetime/date-selection-model.ts

```diff
@@ -38,7 +36,7 @@ export interface DateSelectionModelChange<S> {
 }
 
 /** A selection model containing a date selection. */
-export abstract class MatDateSelectionModel<S, D = ExtractDateTypeFromSelection<S>>
+export abstract class MatDateSelectionModel<D, S = D>
```

Comment:
I originally didn't want to do it this way because I was thinking that end users might need to interact with the selection model and I wanted them to be able to write:

```ts
let selectionModel: MatDateSelectionModel<DateRange<D>>;
```

Instead of:

```ts
let selectionModel: MatDateSelectionModel<D, DateRange<D>>;
```

If we're now treating the selection model as something completely internal to the datepicker that users don't have to mess with, then I'm fine doing it this way.